Sources/REventLogger/DataStorage/REventsStorage.swift

```diff
@@ -0,0 +1,83 @@
+import Foundation
+
+#if SWIFT_PACKAGE
+import class RSDKUtilsMain.AtomicGetSet
+#else
+import class RSDKUtils.AtomicGetSet
+#endif
+
+internal protocol EventDataCacheable {
+
+    func getAllEvents() -> [REvent]
+    func retriveEvent(_ id: String) -> REvent?
+    func getEventCount() -> Int
+    func insertOrUpdateEvent(_ id: String, event: REvent)
+    func deleteEvents(_ ids: [String])
+    func deleteOldEvents(maxCapacity: Int)
+}
+
+internal class EventDataCache: EventDataCacheable {
+
+    private typealias CacheContainer = [String: REvent]
+    private let userDefaults: UserDefaults
+    @AtomicGetSet private var cachedContainer: CacheContainer
+    private let persistedDataKey = "Event_logger_cache"
+
+    init(userDefaults: UserDefaults) {
+        self.userDefaults = userDefaults
+
+        if let persistedData = userDefaults.object(forKey: persistedDataKey) as? Data {
+            do {
+                let decodedData = try JSONDecoder().decode(CacheContainer.self, from: persistedData)
+                cachedContainer = decodedData
+            } catch {
+                cachedContainer = [:]
+            }
+        } else {
+            cachedContainer = [:]
+        }
+    }
+
+    func getAllEvents() -> [REvent] {
+        Array(cachedContainer.values)
+    }
+
+    func retriveEvent(_ id: String) -> REvent? {
+        cachedContainer[id]
+    }
+
+    func deleteEvents(_ ids: [String]) {
+        for key in ids {
+            cachedContainer.removeValue(forKey: key)
+        }
+        saveData()
+    }
+
+    func getEventCount() -> Int {
+        cachedContainer.count
+    }
+
+    func insertOrUpdateEvent(_ id: String, event: REvent) {
+        cachedContainer[id] = event
```

Comment:
what about update event for duplicate events? We can name it as `updateEvent`

Reply:
Duplicate event will checked before this method is called . The type of the will be changed from critical to warning or the occurrence count will be increased then this method will be called to insert the event to DB